lib/PuppeteerSharp.Tests/Network/NetworkEventTests.cs

```diff
@@ -0,0 +1,152 @@
+        {
+            Response response = null;
+            Page.ResponseCreated += (sender, e) => response = e.Response;
+            await Page.GoToAsync(TestConstants.ServerUrl + "/simple.json");
+            Assert.NotNull(response);
+            Assert.Equal("{\"foo\": \"bar\"}\r\n", await response.TextAsync());
+            Assert.Equal(JObject.Parse("{\"foo\": \"bar\"}\r\n"), await response.JsonAsync());
+        }
+
+        [Fact]
+        public async Task PageEventsResponseShouldNotReportBodyUnlessRequestIsFinished()
+        {
+            await Page.GoToAsync(TestConstants.EmptyPage);
+            // Setup server to trap request.
+            var serverResponseCompletion = new TaskCompletionSource<bool>();
+            HttpResponse serverResponse = null;
+            Server.SetRoute("/get", context =>
+            {
+                serverResponse = context.Response;
+                context.Response.WriteAsync("hello ");
+                return serverResponseCompletion.Task;
+            });
+            // Setup page to trap response.
+            Response pageResponse = null;
+            var requestFinished = false;
+            Page.ResponseCreated += (sender, e) => pageResponse = e.Response;
+            Page.RequestFinished += (sender, e) => requestFinished = true;
+            // send request and wait for server response
+            Task WaitForPageResponseEvent()
+            {
+                var completion = new TaskCompletionSource<bool>();
+                Page.ResponseCreated += (sender, e) => completion.SetResult(true);
+                return completion.Task;
+            }
+            await Task.WhenAll(
+                Page.EvaluateExpressionAsync("fetch('/get', { method: 'GET'})"),
+                WaitForPageResponseEvent()
+            );
+
+            Assert.NotNull(serverResponse);
+            Assert.NotNull(pageResponse);
+            Assert.Equal(HttpStatusCode.OK, pageResponse.Status);
+            Assert.False(requestFinished);
+
+            var responseText = pageResponse.TextAsync();
+            // Write part of the response and wait for it to be flushed.
+            await serverResponse.WriteAsync("wor");
+            // Finish response.
+            await serverResponse.WriteAsync("ld!");
+            serverResponseCompletion.SetResult(true);
+            Assert.Equal("hello world!", await responseText);
+        }
+
+        [Fact]
+        public async Task PageEventsRequestFailed()
+        {
+            await Page.SetRequestInterceptionAsync(true);
+            Page.RequestCreated += async (sender, e) =>
+            {
+                if (e.Request.Url.EndsWith("css"))
+                    await e.Request.Abort();
+                else
+                    await e.Request.Continue();
+            };
+            var failedRequests = new List<Request>();
+            Page.RequestFailed += (sender, e) => failedRequests.Add(e.Request);
```

Comment:
RequestFailed is not being raised 😭 in `NetworkManager.OnLoadingFailed` the `_requestIdToRequest` does not contain the requestId

Reply:
@kblok this is the last thing